satpy/readers/__init__.py

```diff
@@ -61,182 +60,101 @@ def _wl_dist(wl_a, wl_b):
-                          resolution and calibration this is the exact
-                          desired list of modifiers for one dataset, not
-                          a list of possible modifiers.
+            polarization (str or list): Dataset polarization
+                                        (ex.'V'). This can also be a
+                                        list of these strings.
+            level (number or list): Dataset level (ex. 100). This can also be a
+                                    list of these numbers.
+            modifiers (list): Modifiers applied to the dataset. Unlike
+                            resolution and calibration this is the exact
+                            desired list of modifiers for one dataset, not
+                            a list of possible modifiers.
 
 
-    Returns (list or DatasetID): Matching key(s)
+    Returns (list or DataID): Matching key(s)
 
     Raises: KeyError if no matching results or if more than one result is
             found when `num_results` is `1`.
 
     """
-    if isinstance(key, numbers.Number):
-        # we want this ID to act as a query so we set modifiers to None
-        # meaning "we don't care how many modifiers it has".
-        key = DatasetID(wavelength=key, modifiers=None)
-    elif isinstance(key, str):
-        # ID should act as a query (see wl comment above)
-        key = DatasetID(name=key, modifiers=None)
-    elif not isinstance(key, DatasetID):
-        raise ValueError("Expected 'DatasetID', str, or number dict key, "
-                         "not {}".format(str(type(key))))
-
-    res = filter_keys_by_dataset_id(key, key_container)
-
-    # further filter by other parameters
-    if resolution is not None:
-        if not isinstance(resolution, (list, tuple)):
-            resolution = (resolution, )
-        res = [k for k in res
-               if k.resolution is not None and k.resolution in resolution]
-    if polarization is not None:
-        if not isinstance(polarization, (list, tuple)):
-            polarization = (polarization, )
-        res = [k for k in res
-               if k.polarization is not None and k.polarization in
-               polarization]
-    if calibration is not None:
-        if not isinstance(calibration, (list, tuple)):
-            calibration = (calibration, )
-        res = [k for k in res
-               if k.calibration is not None and k.calibration in calibration]
-    if level is not None:
-        if not isinstance(level, (list, tuple)):
-            level = (level, )
-        res = [k for k in res
-               if k.level is not None and k.level in level]
-    if modifiers is not None:
-        res = [k for k in res
-               if k.modifiers is not None and k.modifiers == modifiers]
+    key = create_filtered_query(key, query)
+
+    res = filter_keys_by_dataset_query(key, key_container)
+    if not res:
+        raise(KeyError)
```

Comment:
This should be

`raise KeyError("something useful here")`

Reply:
Thanks, fixed in principle